### PR TITLE
Add Endpoint for Commenting on News Articles

### PR DESCRIPTION
**Pull request by bojana-petroska**
**Update:** Implemented the changes after the code review and these are some of the test cases:

Screenshot 2024-12-12 at 11 43 16

Screenshot 2024-12-12 at 11 43 39

Screenshot 2024-12-12 at 11 43 55

Screenshot 2024-12-12 at 12 02 46

**Description:** 

This pull request implements an endpoint for users to comment on news articles. The endpoint allows authenticated users to post a comment on a news article, saving the comment and associated news article in the database. If the news article does not exist in the database, it is created automatically before associating the comment with it.

**Changes:**

1. Added an entity for `Comment` to represent the comment data structure in the database.
2. Created a `route` for comments to handle HTTP requests related to commenting functionality.
3. Developed a `controller` to handle the logic for adding comments on news articles.
4. Added `validation` for the request body using `Joi` to ensure data integrity to required structure.
5. Registered a new main route `(/comments)` in the application to handle all comment-related operations.

**Screenshot:** Below is a screenshot of a successful POST request in Postman, demonstrating the expected behavior of the endpoint:

Screenshot 2024-12-11 at 13 17 14

____

**How to Test:**

*Prerequisites:*

- Register and authenticate a user to obtain a valid token.
- Use Postman or another API client to test the endpoint.
- Include the token in the `Authorization` header: `Bearer <your_token>`.

*Steps:*

Send a `POST` request to the `/comments` endpoint with the following body structure:

```
{
  "title": "News Article Title",
  "description": "Brief description of the article",
  "source": "Article Source",
  "url": "https://example.com/article",
  "publishedAt": "2024-12-10T12:00:00Z",
  "content": "This is my comment"
}
```

*Expected Behavior:*

If the request is valid:

- The news article is created in the database if it does not exist.
- The comment is saved and associated with the news article and the user.
- The response includes the content of the comment and the response message, e.g.:

```
{
  "content": "This is my comment", message: "Comment added successfully"
}
```

*If the request is invalid or the user is not authenticated:*
*Appropriate error messages are returned with corresponding status codes.*

_____

**Please review the implementation and test the functionality. Feedback is welcome!**

### CODE REVIEW

File: src/modules/Comment/controller.ts [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+        description,
+        title,
+      });
+      await newsRepository.save(newsArticle);
+    }
+
+    const comment = commentRepository.create({
+      content,
+      timeStamp: timeStamp ? new Date(timeStamp) : null,
+      user,
```
Comment (MariemSo) [inline src/modules/Comment/controller.ts:40]: save only userId

File: src/modules/Comment/controller.ts [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+      news: newsArticle,
+    });
+
+    await commentRepository.save(comment);
+
+    res.status(201).send({
+      content: comment.content,
```
Comment (MariemSo) [inline src/modules/Comment/controller.ts:47]: save the full comment, so it is possible to retrieve everything related to it

File: src/modules/Comment/controller.ts [hunk 3]
```diff
@@ -0,0 +1,58 @@
+import type { Request, Response } from 'express';
+import { AppDataSource } from '../../config/database.js';
+import { News } from '../../entities/News.js';
+import { Comment } from '../../entities/Comment.js';
+
+const commentRepository = AppDataSource.getRepository(Comment);
+const newsRepository = AppDataSource.getRepository(News);
+
+const commentOnNewsArticle = async (req: Request, res: Response) => {
+  const { user } = req;
```
Comment (MariemSo) [inline src/modules/Comment/controller.ts:10]: get the userId from req.user

**Review by Francieli-S — CHANGES_REQUESTED** [review]
Nice job, Bojana! The code is clean and easy to read. I also really liked how you wrote the PR description, which iss very detailed and well-structured.

File: src/modules/Comment/controller.ts (hunk 2 above)
Comment (Francieli-S) [reply src/modules/Comment/controller.ts:47]: The full comment is saved in line 44. And just the comment.content send in the response.

Comment (bojana-petroska) [reply src/modules/Comment/controller.ts:47]: @MariemSo @Francieli-S yes, the full comment is saved, but not fully sent to the client. and just need to be edited to same only the userId and not the whole user.

**Comment by bojana-petroska** [comment]
Thanks @MariemSo & @Francieli-S for the feedback! Highly appreciate it!

**Review by Francieli-S — CHANGES_REQUESTED** [review]

File: src/modules/Comment/controller.ts [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+        title,
+      });
+      await newsRepository.save(newsArticle);
+    }
+
+    const comment = commentRepository.create({
+      content,
+      timeStamp: timeStamp ? new Date(timeStamp) : null,
+      user,
+      news: newsArticle,
```
[the remark this answers is not in the document]
Comment (bojana-petroska) [reply src/modules/Comment/controller.ts:41]: yes, makes sense. I adjusted it, thanks!

File: src/modules/Comment/validation.ts [hunk 5]
```diff
@@ -0,0 +1,14 @@
+import Joi from 'joi';
+
+const comments = {
+  body: Joi.object({
+    title: Joi.string().required(),
+    description: Joi.string().required(),
+    source: Joi.string().required(),
+    url: Joi.string().required(),
```
Comment (regikov) [inline src/modules/Comment/validation.ts:8]: hey @bojana-petroska i would also add that url should be valid. By adding .uri() -> this validates  that the string is a valid URI

**Review by ahzia — APPROVED** [review]
Good Job @bojana-petroska 👏👏👏
There is no changes required, I will merge your branch.

File: .vscode/settings.json
Comment (ahzia) [inline .vscode/settings.json:1]: Can you add back this file?

File: docker-compose.yml
Comment (ahzia) [inline docker-compose.yml:1]: Can you add back this file?

**Pull request merged**